src/connection.cpp

```diff
@@ -191,15 +217,106 @@ namespace oxen::quic
         return "{:02x}"_format(fmt::join(std::begin(data), std::begin(data) + datalen, ""));
     }
 
-    uint64_t Connection::next_dgram_id()
+    void Connection::packet_io_ready()
     {
-        _last_dgram_id += _increment;
-        return _last_dgram_id;
+        event_active(packet_io_trigger.get(), 0, 0);
     }
 
-    void Connection::packet_io_ready()
+    void Connection::handle_conn_packet(const Packet& pkt)
     {
-        event_active(packet_io_trigger.get(), 0, 0);
+        if (auto rv = ngtcp2_conn_in_closing_period(*this); rv != 0)
+        {
+            log::trace(log_cat, "Note: CID-{} in closing period; signaling endpoint to delete connection", scid());
+
+            _endpoint.call([&]() {
+                log::debug(
+                        log_cat, "Error: connection (CID: {}) is in closing period; endpoint deleting connection", scid());
+                _endpoint.delete_connection(scid());
+            });
+            return;
+        }
+
+        if (is_draining())
+        {
+            log::debug(log_cat, "Error: connection is already draining; dropping");
+        }
+
+        // TODO: if read packet gives us failure, should we close?
+        if (read_packet(pkt).success())
+            log::trace(log_cat, "done with incoming packet");
+        else
+            log::trace(log_cat, "read packet failed");  // error will be already logged
+    }
+
+    io_result Connection::read_packet(const Packet& pkt)
+    {
+        auto ts = get_timestamp().count();
+        log::trace(log_cat, "Calling ngtcp2_conn_read_pkt...");
+        auto rv = ngtcp2_conn_read_pkt(*this, pkt.path, &pkt.pkt_info, u8data(pkt.data), pkt.data.size(), ts);
+
+        switch (rv)
+        {
+            case 0:
+                packet_io_ready();
+                break;
+            case NGTCP2_ERR_DRAINING:
+                log::trace(log_cat, "Note: CID-{} is draining; signaling endpoint to drain connection", scid());
+                _endpoint.call([&]() {
```

Comment:
These (this one and the various ones that follow) could also be `[this]` captures